wave_lang/kernel/wave/codegen/emitter.py

```diff
@@ -226,6 +234,220 @@ def emit(self, graph: Optional[fx.Graph] = None) -> Operation:
+            if b.binding_type not in (BindingType.SYMBOL_VALUE, BindingType.INDEX_VALUE)
+        ]
+        for i, b in enumerate(host_bindings):
+            binding_map[id(b)] = i
+            if b.binding_type == BindingType.KERNEL_BUFFER:
+                for j, symbol in enumerate(b.kernel_buffer_type.symbolic_shape):
+                    if symbol in symbol_map:
+                        continue
+                    symbol_map[symbol] = (i, j)
+
+        bindings = self.root_sig.sig.linear_bindings
+
+        ptr = llvm_d.PointerType.get()
+
+        def abi_type(binding: BindingDesc):
+            if binding.binding_type == BindingType.KERNEL_BUFFER:
+                # Buffer passed to kernel as 0D memrefs to simplify ABI.
+                element_type = IrType.parse(
+                    binding.kernel_buffer_type.dtype.ir_type_asm()
+                )
+                return MemRefType.get([], element_type=element_type)
+
+            # Scalars are passed as is.
+            return binding.as_mlir_type()
+
+        arg_types = [abi_type(b) for b in bindings]
+
+        ftype = FunctionType.get(arg_types, [])
+        locs = [a.location for a in kernel_func.body.blocks[0].arguments]
+
+        gpu_module = gpu_d.GPUModuleOp("gpu_module")
+        gpu_module.parent.operation.attributes["gpu.container_module"] = UnitAttr.get()
+        module_block = gpu_module.bodyRegion.blocks.append()
+
+        with InsertionPoint(module_block), Location.name("wave-generated gpu module"):
+            kernel_func_wrapper = gpu_d.GPUFuncOp(
+                TypeAttr.get(ftype), sym_name=self.kernel_name, kernel=True
+            )
+
+        kernel_entry_block = kernel_func_wrapper.body.blocks.append(
+            *arg_types,
+            arg_locs=locs,
+        )
+
+        # Inline the kernel function into the gpu module function body and erase the original function
+        with (
+            InsertionPoint(kernel_entry_block),
+            Location.name("wave-generated kernel function"),
+        ):
+            # Move operations except terminator
+            ops_to_move = list(kernel_func.entry_block)[:-1]
+            for op in ops_to_move:
+                op.detach_from_parent()
+                kernel_entry_block.append(op)
+
+            # Replace all uses of old arguments
+            for old_arg, new_value in zip(
+                kernel_func.entry_block.arguments, kernel_entry_block.arguments
+            ):
+                old_arg.replace_all_uses_with(new_value)
+
+            gpu_d.return_([])
+            kernel_func.erase()
+
+        # Declare runtime functions
+        i32 = IntegerType.get_signless(32)
+        i64 = IntegerType.get_signless(64)
+        f64 = F64Type.get()
+
+        buffer_type = MemRefType.get(
+            [MemRefType.get_dynamic_size()], element_type=IntegerType.get_signless(8)
+        )
+
+        # wave_get_buffer returns memref so we need emit_c_interface to have a C-compatible ABI
```

Comment:
I'd just make all of them have the interface then. This also changes the name of the function to call and generally having a special case leads to more complexity and harder time debugging this. Unless there's a noticeable performance hit in doing so (I expect LLVM to inline these anyway).